src/hal/motor.cpp

```diff
@@ -2,20 +2,461 @@
+    if (digitalRead(PIN_LIMIT_TOP) == PIN_LIMIT_TRIPPED) {
+      // Do not move the motor in the OPEN direction if the top limit switch is tripped.
+      set_motor_state(MOTOR_STATE_HOLD);
+      motor_position_zero();
+    } else {
+      set_motor_state(MOTOR_STATE_OPEN);
+    }
+  } else {
+    if ((motor_state == MOTOR_STATE_OPEN) ||
+        (motor_state == MOTOR_STATE_CLOSE)) {
+      // Hold the motor in place in order to maintain the correct position
+      // without having to zero the motor again.
+      set_motor_state(MOTOR_STATE_HOLD);
+    }
+  }
+}
+
+// Motor Position Tracking ISR
+static inline void motor_position_tracking_ISR()
+{
+  if (motor_state == MOTOR_STATE_OPEN) {
+    if (--motor_control.microstep_position <= -MC_MICROSTEPS_PER_STEP) {
+      motor_control.step_position--;
+      motor_control.microstep_position = 0;
+    }
+  } else if (motor_state == MOTOR_STATE_CLOSE) {
+    if (++motor_control.microstep_position >= MC_MICROSTEPS_PER_STEP) {
+      motor_control.step_position++;
+      motor_control.microstep_position = 0;
+    }
+  }
+}
+
+ISR(TIMER3_COMPA_vect)
+{
+  motor_state_control_ISR();
+}
+
+ISR(TIMER4_COMPA_vect)
+{
+  motor_position_tracking_ISR();
+}
+
+//*****************************************************************************
+// motorHal Interface Implementation
+//*****************************************************************************
 int motorHalInit(void)
 {
-  // TODO: Implement
+  uint8_t sreg = SREG; // save interrupt state
+  cli(); // disable interrupts
+
+  timer3_setup();
+  timer4_setup();
+
+  // Serial Debug Port
+  // Serial.begin(38400);
+  // Serial.print("Start");
+  // Serial.println();
+
+  // Motor Pin Configuration
+  pinMode(PIN_MOTOR_PWM, OUTPUT);
+  pinMode(PIN_MOTOR_ENABLE, OUTPUT);
+  pinMode(PIN_MOTOR_DIRECTION, OUTPUT);
+
+  // Limit Switch Pin Configuration
+  pinMode(PIN_LIMIT_BOTTOM, INPUT_PULLUP);
+  pinMode(PIN_LIMIT_TOP, INPUT_PULLUP);
+
+  SREG = sreg; // restore interrupt state
+
+  // Move the motor to the home position to zero the motor position.
+  motor_home();
+  while (motor_busy) {}
+  delay(1000);
+
+  // Move to the top of the bag.
+  // TODO: This value will vary depending upon the installation location of the
+  // top limit switch, as well as the type of bag. Ultimately, the sensors
+  // should be used to find the top of the bag.
+  motor_move(15, MOTOR_DIRECTION_CLOSE);
+  while (motor_busy) {} 
+  delay(1000);
+
   return HAL_OK;
 }
 
+// TODO: Implement duration.
 int motorHalBegin(unsigned int direction, unsigned int distance, unsigned int duration)
 {
-  // TODO: Implement
+  motor_move_command.distance = distance;
+
+  if (direction == MOTOR_HAL_DIRECTION_INHALATION) {
+    motor_move_command.direction = MOTOR_DIRECTION_CLOSE;
+  } else if (direction == MOTOR_HAL_DIRECTION_EXHALATION) {
+    motor_move_command.direction = MOTOR_DIRECTION_OPEN;
+  } else {
+    return HAL_FAIL;
+  }
+
+  motor_move_command.valid = true;
```

Comment:
I didn't realize that *HalBegin() was allowed to initiate the operation.  (Based on the naming convention, I was under the impression that it would just set up the command to be started when *HalRun is called.) If this is the case, I'd suggest that we consider an alternate naming convention, such as *HalBegin() and *HalStatus().

Reply:
So, in theory, if your motor driver did need work done, you would execute it in that `motorHalRun`, which is why it named that (like, `motorHalBegin` would set everything up and begin the operation, `motorHalRun` would run the maybe the FSM depending on the implementation) since it would be expected to be called frequently. However, we can rename it if we feel like this is no longer is a valid assumption.

Reply:
It's probably not a good idea for a device driver to have a built-in expectation that one of it's functions should be called at some regular interval in order to function properly.

Since the sensor drivers also don't have this built-in assumption, I'd like to change the interface naming convention to make it more clear.

Reply:
They actually used to in an earlier iteration (or at least that they would be called with at least some frequency, not a precise frequency); but please feel free to refactor

Reply:
I've decided to defer the interface changes to a later commit.